services/src/main/scala/cromwell/services/MetadataServiceActor.scala

```diff
@@ -34,3 +37,25 @@ object MetadataServiceActor {
   }
   */
 }
+
+object MetadataServiceActorImplicits {
+  implicit class EnhancedServiceRegistryActorForMetadata(val actor: ActorRef) extends AnyVal {
+    def pushWdlValueMetadata(metadataKey: MetadataKey, output: WdlValue): Unit = output match {
+      case WdlArray(_, valueSeq) =>
+        val zippedSeq = valueSeq.zipWithIndex
+        zippedSeq foreach { case (value, index) => actor.pushWdlValueMetadata(metadataKey.copy(key = s"${metadataKey.key}[$index]"), value) }
+      case WdlMap(_, valueMap) =>
+        valueMap foreach { case (key, value) => actor.pushWdlValueMetadata(metadataKey.copy(key = metadataKey.key + s":$key"), value) }
+      case value =>
+        actor ! PutMetadataAction(MetadataEvent(metadataKey, MetadataValue(value.valueString)))
+    }
+
+    def pushThrowableMetadata(metadataKey: MetadataKey, t: Throwable): Unit = {
+      actor ! PutMetadataAction(MetadataEvent(metadataKey.copy(key = s"${metadataKey.key}:message"), MetadataValue(t.getMessage)))
+      t.getStackTrace.zipWithIndex foreach { case (trace, index) =>
+        actor ! PutMetadataAction(MetadataEvent(metadataKey.copy(key = s"${metadataKey.key}:stackTrace[$index]"), MetadataValue(trace.toString)))
```

Comment:
Long PR discussion is long. This might be best discussed face to face. I'll make it a tech-talk topic for when everyone's back after the break.

Worst case... we'll just need a ticket to reverse this :)